### PR TITLE
FIX: When creating new message via URL do not redirect

### PR DESCRIPTION
**Pull request by arpitjalan**
If a user clicks on `/new-message` route from inside the instance we're redirecting the user to `/latest` page which is only intended if the user is coming from an external site. This commit checks for this condition and only redirects when user is coming from external source.

This also makes the behavior consistent with `new-topic` route.

Internal topic reference: `/t/-/129523/`

### CODE REVIEW

File: app/assets/javascripts/discourse/app/routes/new-message.js [hunk 1]
```diff
@@ -12,49 +12,69 @@ export default DiscourseRoute.extend({
...
+                      I18n.t("composer.cant_send_pm", { username: groupName })
+                    );
+                  }
+                })
+                .catch(() => this.dialog.alert(I18n.t("generic_error")));
+            } else {
+              this.openComposer(transition);
+            }
+          });
+      }
```
Comment (Grubba27) [inline app/assets/javascripts/discourse/app/routes/new-message.js:64]: It is hard to follow this component, and I felt overwhelmed when I first looked at it.

Moving it to an early return pattern instead of those else if branches would not make it easier to understand and maintain? 

this is this same piece _without_ the changes but with this refactor
```js

    if (!this.currentUser) {
      cookie("destination_url", window.location.href);
      this.router.replaceWith("login");
      return;
    }

    this.router
      .replaceWith("discovery.latest")
      .followRedirects()
      .then(() => {
        if (params.username) {
          return this.composer.openNewMessage({
            recipients: params.username,
            title: params.title,
            body: params.body,
          });
        }

        if (groupName) {
          return Group.messageable(groupName).then((result) => {
            if (!result.messageable) {
              return this.dialog.alert(
                I18n.t("composer.cant_send_pm", { username: groupName })
              );
            }
            return next(() =>
              this.composer.openNewMessage({
                recipients: groupName,
                title: params.title,
                body: params.body,
              })
            );
          });
        }
        return this.composer.openNewMessage({
          title: params.title,
          body: params.body,
        });
      });

```

Comment (arpitjalan) [reply app/assets/javascripts/discourse/app/routes/new-message.js:64]: Good idea @Grubba27, updated the PR to prefer early return pattern.

File: app/assets/javascripts/discourse/app/routes/new-message.js [hunk 2]
```diff
@@ -12,49 +12,74 @@ export default DiscourseRoute.extend({
...
+          .then((result) => {
+            if (result.messageable) {
+              this.openComposer(transition, groupName);
+            } else {
+              this.dialog.alert(
+                I18n.t("composer.cant_send_pm", { username: groupName })
+              );
+            }
+          })
+          .catch(() => this.dialog.alert(I18n.t("generic_error")));
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/routes/new-message.js:43]: I know you just moved code around, but any chances we could provide the user with a better error message than a generic one? 🤔

**Review by ZogStriP — APPROVED** [review]

File: app/assets/javascripts/discourse/app/routes/new-message.js (hunk 2 above)
Comment (arpitjalan) [reply app/assets/javascripts/discourse/app/routes/new-message.js:43]: Should we just pass along the error message then? Not sure what specifics we can provide here because even we're not sure what's the error in this case.

Note just above we're giving clear error message `I18n.t("composer.cant_send_pm", { username: groupName })` because we know that the group is not messageable, but in this case this just might be network error, hence we're falling back to generic error.

Maybe a better worded message like: "Sorry, there was an error creating that message. Please try again."?

Comment (ZogStriP) [reply app/assets/javascripts/discourse/app/routes/new-message.js:43]: If we truly don't know what happened, then sure, improving the generic message could help 👍
